File: packages/engine/src/memory/taskMemory.ts
import { taskMemory as dbTaskMemory, getDrizzle } from 'db';
import type { SubTask } from 'types';

export type { SubTask };

export interface TaskState {
  goal: string;
  subtasks: SubTask[];
  currentStepIndex: number;
}

export class TaskMemory {
  private threadId: string;

  constructor(threadId: string) {
    this.threadId = threadId;
  }

  async getTaskState(): Promise<TaskState | null> {
    const dbInstance = getDrizzle();
    if (dbInstance) {
      try {
        const { eq } = require('drizzle-orm');
        const rows = await dbInstance
          .select()
          .from(dbTaskMemory)
          .where(eq(dbTaskMemory.threadId, this.threadId))
          .limit(1);
        if (rows && rows.length > 0) {
          const record = rows[0];
          return record.pendingIntents as TaskState;
        }
      } catch (err) {
        console.warn('[TaskMemory] Failed to get task state from DB:', err);
      }
    }
    return null;
  }

  async saveTaskState(state: TaskState): Promise<void> {
    console.log(`[TaskMemory] Saving state for thread ${this.threadId}: ${JSON.stringify(state)}`);
    const dbInstance = getDrizzle();
    if (dbInstance) {
      try {
        const { eq } = require('drizzle-orm');
        // Check if there is an existing record
        const rows = await dbInstance
          .select()
          .from(dbTaskMemory)
          .where(eq(dbTaskMemory.threadId, this.threadId))
          .limit(1);

        if (rows && rows.length > 0) {
          // Update existing task state
          await dbInstance
            .update(dbTaskMemory)
            .set({
              pendingIntents: state,
              updatedAt: new Date(),
            })
            .where(eq(dbTaskMemory.threadId, this.threadId));
        } else {
          // Insert new task state
          await dbInstance.insert(dbTaskMemory).values({
            threadId: this.threadId,
            pendingIntents: state,
            updatedAt: new Date(),
          });
        }
        console.log(`[TaskMemory] Successfully persisted state to DB for thread ${this.threadId}`);
      } catch (err) {
        console.warn('[TaskMemory] Failed to save task state to DB:', err);
      }
    }
  }
}

File: services/engine-py/src/engine_py/memory/task_memory.py
"""任务记忆 — 镜像 packages/engine/src/memory/taskMemory.ts(task_memory 表 upsert)。"""

from __future__ import annotations

import datetime as _dt

from sqlalchemy import select

from ..db import TaskMemoryRow, get_session


class TaskMemory:
    def __init__(self, thread_id: str) -> None:
        self.thread_id = thread_id

    async def get_task_state(self) -> dict | None:
        try:
            async with get_session() as session:
                row = (
                    await session.execute(
                        select(TaskMemoryRow).where(TaskMemoryRow.thread_id == self.thread_id).limit(1)
                    )
                ).scalar_one_or_none()
                return row.pending_intents if row else None
        except Exception as err:
            print(f"[TaskMemory] Failed to get task state from DB: {err}")
            return None

    async def save_task_state(self, state: dict) -> None:
        try:
            async with get_session() as session:
                row = (
                    await session.execute(
                        select(TaskMemoryRow).where(TaskMemoryRow.thread_id == self.thread_id).limit(1)
                    )
                ).scalar_one_or_none()
                if row:
                    row.pending_intents = state
                    row.updated_at = _dt.datetime.now()
                else:
                    session.add(
                        TaskMemoryRow(
                            thread_id=self.thread_id,
                            pending_intents=state,
                            updated_at=_dt.datetime.now(),
                        )
                    )
                await session.commit()
        except Exception as err:
            print(f"[TaskMemory] Failed to save task state to DB: {err}")
